### PR TITLE
DOC-749 Update Creating a Plugin page

### PR DESCRIPTION
**Pull request by metricjs**
Related Ticket: DOC-749

Description of Changes:
* Updated the Creating a Plugin page with newer recommendations and more details.

Pre-checks:
- [x] Branch prefixed with `feature/` or `hotfix/`
- [x] `_data/nav.yml` has been updated (if applicable) - N/A
- [x] Files has been included where required (if applicable) - N/A
- [x] Files removed have been deleted, not just excluded from the build (if applicable) - N/A
- [x] (New product features only) Release Note added - N/A

Review:
- [x] Documentation Team Lead has reviewed
- [x] Product Manager has reviewed

### CODE REVIEW

File: advanced/creating-a-plugin.md [hunk 1]
```diff
@@ -1,43 +1,113 @@
...
 
-The build scripts that comes with the {{site.productname}} development package automatically builds all plugins including custom ones.
+{{site.productname}} does not require any special file structure or tooling apart from these requirements, so custom plugins can be developed using most frameworks and tools. 
 
-## Example of the plugin file structure
+### Yeoman Generator
 
+{{site.companyname}} maintains a [Yeoman generator]({{ site.baseurl }}/advanced/yeoman-generator/) to help make plugin creation for {{site.productname}} easier. The Yeoman Generator will create the files and boilerplate code required for a custom plugin, and sets up some helpful commands.
```
Comment (JamesToohey) [inline advanced/creating-a-plugin.md:20]: Thoughts?

```suggestion
{{site.companyname}} maintains a [Yeoman generator]({{ site.baseurl }}/advanced/yeoman-generator/) to assist with creating plugins for {{site.productname}}. The Yeoman Generator will create the files and boilerplate code required for a custom plugin, and sets up some helpful commands. 
```

File: advanced/creating-a-plugin.md [hunk 2]
```diff
@@ -1,43 +1,113 @@
...
-The build scripts that comes with the {{site.productname}} development package automatically builds all plugins including custom ones.
+{{site.productname}} does not require any special file structure or tooling apart from these requirements, so custom plugins can be developed using most frameworks and tools. 
 
-## Example of the plugin file structure
+### Yeoman Generator
 
+{{site.companyname}} maintains a [Yeoman generator]({{ site.baseurl }}/advanced/yeoman-generator/) to help make plugin creation for {{site.productname}} easier. The Yeoman Generator will create the files and boilerplate code required for a custom plugin, and sets up some helpful commands. 
+
+## Registering a custom plugin with {{site.productname}}
+Register a custom plugin with {{site.productname}} using the PluginManager. `PluginManager.add()` takes a string for the plugin identifier and a function that contains the code for initializing the plugin. Optionally return an object containing a function that returns metadata which will be used in the list of plugins in the [Help plugin]({{ site.baseurl }}/plugins/opensource/help) dialog.
```
Comment (JamesToohey) [inline advanced/creating-a-plugin.md:23]: Finding this part a bit hard to read

```
Optionally return an object containing a function that returns metadata which will be used in the list of plugins in the [Help plugin]({{ site.baseurl }}/plugins/opensource/help) dialog.
```

File: advanced/creating-a-plugin.md [hunk 3]
```diff
@@ -1,43 +1,113 @@
...
+
+The plugin identifier passed to `PluginManager.add()` is used by {{site.productname}} as an identifier string. It should be alphanumeric and can not contain spaces. Plugin identifiers should be unique. If multiple plugins have the same identifier, one will override the others.
```
Comment (JamesToohey) [inline advanced/creating-a-plugin.md:25]:
```suggestion
The plugin identifier passed to `PluginManager.add()` is used by {{site.productname}} as an identifier string. It should be alphanumeric and cannot contain spaces. Plugin identifiers should be unique. If multiple plugins have the same identifier, one will override the others. 
```

File: advanced/creating-a-plugin.md [hunk 4]
```diff
@@ -1,43 +1,113 @@
...
 
-Developers can designate the location of the plugins by loading the `plugin.js` or `plugin.min.js` files directly after the `tinymce.js` or `tinymce.min.js`.
+## Testing {{site.productname}}
+
+Due to the amount of browser APIs used by {{site.productname}}, when testing {{site.productname}} or any custom plugins, you must use a testing framework that runs on a real browser. Testing frameworks that mock the browser will not work. 
+
+## Language localization
 
-## Example of loading the plugin from another URL
+{{site.productname}} comes with translations for many languages. To add additional strings to a supported language for a custom plugin, use the following procedure.
```
Comment (JamesToohey) [inline advanced/creating-a-plugin.md:87]:
```suggestion
{{site.productname}} comes with translations for many languages. To add additional strings to a supported language for a custom plugin, use the following procedure:
```

File: advanced/yeoman-generator.md [hunk 5]
```diff
@@ -7,7 +7,7 @@ description: How to use the Yeoman generator to bootstrap a new TinyMCE plugin u
 keywords: webpack yeoman generator plugin tinymce
 ---
 
-The {{site.productname}} Plugin Generator is designed to make it quick and easy to get started creating plugins to extend and enhance your rich-text editing experience.
+{{site.companyname}} maintains a [Yeoman generator]({{ site.baseurl }}/advanced/yeoman-generator/) to help make plugin creation for {{site.productname}} easier. The Yeoman Generator will create the files and boilerplate code required for a custom plugin, and sets up some helpful commands.
```
Comment (JamesToohey) [inline advanced/yeoman-generator.md:10]:
```suggestion
{{site.companyname}} maintains a [Yeoman generator]({{ site.baseurl }}/advanced/yeoman-generator/) to assist with creating plugins for {{site.productname}}. The Yeoman Generator will create the files and boilerplate code required for a custom plugin, and sets up some helpful commands. 
```

**Review by JamesToohey — COMMENTED** [review]
Just a quick skim with some thoughts and minor tweaks

File: advanced/creating-a-plugin.md [hunk 6]
```diff
@@ -1,43 +1,113 @@
...
+{{site.productname}} is designed to be easily extended by custom plugins.
 
-## File Structure
+## Requirements 
 
-First create a directory in the {{site.productname}} plugins directory. {{site.productname}} loads the `plugin.js` file when developers use the `tinymce.js` file in the page. Similarly, the `tinymce.min.js` loads the `plugin.min.js` file. The recommended usage is to use the `tinymce.js` file while developing and then use a build script minifying the `plugin.js` into `plugin.min.js` in production.
+To be recognised as a plugin by {{site.productname}}, the code for your custom plugin must have a JavaScript file with a single entry point that registers the plugin with {{site.productname}} using the `PluginManager`. Any other code or resources can be in separate files and can be loaded in any standard manner. {{site.productname}} also has various APIs for loading [scripts]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom.scriptloader/) and [stylesheets]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom).stylesheetloader/).
```
Comment (ashfordneil) [inline advanced/creating-a-plugin.md:14]: Unsure on this, what do you think
```suggestion
To be recognised as a plugin by {{site.productname}}, the code for your custom plugin must have a JavaScript file with a single entry point that registers the plugin with {{site.productname}} using the `PluginManager` API. Any other code or resources can be in separate files and can be loaded in any standard manner. {{site.productname}} also has various APIs for loading [scripts]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom.scriptloader/) and [stylesheets]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom).stylesheetloader/).
```

**Review by ashfordneil — APPROVED** [review]

File: advanced/creating-a-plugin.md (hunk 3 above)
Comment (metricjs) [reply advanced/creating-a-plugin.md:25]: Not sure which we prefer tbh. @tylerkelly13 ?

File: advanced/creating-a-plugin.md (hunk 4 above)
Comment (metricjs) [reply advanced/creating-a-plugin.md:87]: Hm. @tylerkelly13 ?

File: advanced/creating-a-plugin.md [hunk 7]
```diff
@@ -1,43 +1,113 @@
...
-/tinymce/plugins/example/plugin.min.js
+
+## Using custom plugins with {{site.productname}}
+
+Custom plugins can be used by a {{site.productname}} instance by either:
+
+- **Using `external_plugins`**: Use the [external_plugins option]({{ site.baseurl }}/configure/integration-and-setup/#external_plugins) to specify the URL based location of the entry point file for the plugin.
+- **Copy code into `plugins` folder**: Copy the entry point file (and any other files) into the `plugins` folder of the distributed {{site.productname}} code.
+
+The plugin can then be used by including it in the list of plugins specified by the [`plugins` option]({{ site.baseurl }}/configure/integration-and-setup/#plugins).
```
Comment (tylerkelly13) [inline advanced/creating-a-plugin.md:56]: should this be a part of the point above, or is the example for `external_plugins` wrong (empty `plugins` list)?

File: advanced/creating-a-plugin.md [hunk 8]
```diff
@@ -1,43 +1,113 @@
...
-  plugins: 'example'
+  selector: 'textarea',  // change this value according to your HTML
+  plugins: 'pluginId'
 });
 ```
 
-Developers can designate the location of the plugins by loading the `plugin.js` or `plugin.min.js` files directly after the `tinymce.js` or `tinymce.min.js`.
+## Testing {{site.productname}}
+
+Due to the amount of browser APIs used by {{site.productname}}, when testing {{site.productname}} or any custom plugins, you must use a testing framework that runs on a real browser. Testing frameworks that mock the browser will not work.
```
Comment (tylerkelly13) [inline advanced/creating-a-plugin.md:83]:
```suggestion
Due to the range of browser APIs used by {{site.productname}}; when testing {{site.productname}} or any custom plugins, you must use a testing framework that runs on a real browser. Testing frameworks that mock the browser will not work. 
```

File: advanced/creating-a-plugin.md (hunk 4 above)
Comment (tylerkelly13) [reply advanced/creating-a-plugin.md:87]: I would use the colon personally.

File: advanced/creating-a-plugin.md [hunk 9]
```diff
@@ -1,43 +1,113 @@
...
+
+## Language localization
 
-## Example of loading the plugin from another URL
+{{site.productname}} comes with translations for many languages. To add additional strings to a supported language for a custom plugin, use the following procedure.
 
-```html
-<script src="/tinymce/js/tinymce.min.js"></script>
-<script src="/scripts/my.tinymce.plugin.js"></script>
+1. Create a "langs" directory in the plugin directory for custom translations.
```
Comment (tylerkelly13) [inline advanced/creating-a-plugin.md:89]:
```suggestion
1. Create a "`langs`" directory in the custom plugin's root directory for custom translations. 
```

File: advanced/creating-a-plugin.md (hunk 2 above)
Comment (metricjs) [reply advanced/creating-a-plugin.md:23]: Hm. I could just move this down to where I detail the metadata object...

File: advanced/creating-a-plugin.md [hunk 10]
```diff
@@ -1,43 +1,113 @@
...
+## Language localization
 
-## Example of loading the plugin from another URL
+{{site.productname}} comes with translations for many languages. To add additional strings to a supported language for a custom plugin, use the following procedure.
 
-```html
-<script src="/tinymce/js/tinymce.min.js"></script>
-<script src="/scripts/my.tinymce.plugin.js"></script>
+1. Create a "langs" directory in the plugin directory for custom translations. 
+2. For each language that the plugin supports, create a translation file. The files should be JavaScript files and use the relevant language code as the file name. For example, {{site.productname}} will look for a Spanish translation file at `'<your plugin>/langs/es_ES.js'`, where `<your plugin>` is to the directory that contains the plugin's entry point file. For a list of supported languages, see: [Supported languages]({{ site.baseurl }}/configure/localization/#supportedlanguages).
```
Comment (tylerkelly13) [inline advanced/creating-a-plugin.md:90]:
```suggestion
2. For each language that the plugin supports, create a translation file.

    The files should be JavaScript files and use the relevant language code as the file name. For example: {{site.productname}} will search for a Spanish translation file at `'<your plugin>/langs/es_ES.js'`, where `<your plugin>` is to the directory that contains the plugin's entry point file. For a list of supported languages, see: [Supported languages]({{ site.baseurl }}/configure/localization/#supportedlanguages).
    
```
look = personification, search, less so.
Also, the indentation _shouldn't_ break the numbering make the para a child of the above list item...

File: advanced/creating-a-plugin.md [hunk 11]
```diff
@@ -1,43 +1,113 @@
...
+2. In each translation file, add translation strings by passing an object containing key-value pairs of source strings and translation strings to `tinymce.addI18n()`.
```
Comment (tylerkelly13) [inline advanced/creating-a-plugin.md:91]:
```suggestion
2. In each translation file, add translation strings by passing an object containing key-value pairs of source strings and translation strings to the [`tinymce.addI18n()` API]({{site.baseurl}}/api/tinymce/root_tinymce/#addi18n).
```

File: advanced/creating-a-plugin.md [hunk 12]
```diff
@@ -1,43 +1,113 @@
...
+### Yeoman Generator
 
+{{site.companyname}} maintains a [Yeoman generator]({{ site.baseurl }}/advanced/yeoman-generator/) to help make plugin creation for {{site.productname}} easier. The Yeoman Generator will create the files and boilerplate code required for a custom plugin, and sets up some helpful commands. 
+
+## Registering a custom plugin with {{site.productname}}
+Register a custom plugin with {{site.productname}} using the PluginManager. `PluginManager.add()` takes a string for the plugin identifier and a function that contains the code for initializing the plugin. Optionally return an object containing a function that returns metadata which will be used in the list of plugins in the [Help plugin]({{ site.baseurl }}/plugins/opensource/help) dialog.
+
+The plugin identifier passed to `PluginManager.add()` is used by {{site.productname}} as an identifier string. It should be alphanumeric and can not contain spaces. Plugin identifiers should be unique. If multiple plugins have the same identifier, one will override the others. 
+
+The metadata object can contain the following values:
```
Comment (metricjs) [inline advanced/creating-a-plugin.md:27]:
```suggestion
Optionally, the function passed to `PluginManager.add()` can return an object that contains data that {{ site.productname }} or other plugins can use. {{ site.companyname }} recommend including a `getMetadata` callback that returns an object containing data that can be used to populate the list of plugins in the [Help plugin]({{ site.baseurl }}/plugins/opensource/help) dialog. The metadata object should contain the following values:
```
Maybe this @JamesToohey ?

**Review by tylerkelly13 — CHANGES_REQUESTED** [review]
Definitely better than it was :)

File: advanced/creating-a-plugin.md (hunk 7 above)
Comment (metricjs) [reply advanced/creating-a-plugin.md:56]: Ah, yes, should be moved up. Thanks!

Comment (metricjs) [inline advanced/creating-a-plugin.md:56]:
```suggestion
- **Copy code into `plugins` folder**: Copy the entry point file (and any other files) into the `plugins` folder of the distributed {{site.productname}} code. The plugin can then be used by including it in the list of plugins specified by the [`plugins` option]({{ site.baseurl }}/configure/integration-and-setup/#plugins).
```

File: advanced/creating-a-plugin.md (hunk 12 above)
Comment (metricjs) [reply advanced/creating-a-plugin.md:33]: @tylerkelly13 FYI. I realised this object (the one that's returned, not the metadata one) can contain whatever methods - that's how we do plugin APIs. Hence kind of implying that here. `contains data` should maybe be `contains functions` or `contains callbacks`, not sure. Can show you an example from our plugins if that might help.

Comment (metricjs) [reply advanced/creating-a-plugin.md:33]: ...Theoretically the metadata object can contain more data too, those two fields are just the ones the Help plugin uses...

**Comment by tylerkelly13** [comment]
changed the comments in the example to multi-line comments, single line comments break the demos apparently. I need to look into it.

File: advanced/creating-a-plugin.md (hunk 2 above)
Comment (tylerkelly13) [reply advanced/creating-a-plugin.md:23]: Yep. You're right. It should be with the metadata stuff

File: advanced/creating-a-plugin.md [hunk 13]
```diff
@@ -7,139 +7,118 @@ description: A short introduction to creating plugins for TinyMCE along with an
...
+
+1. Create a "`langs`" directory in the custom plugin's root directory for custom translations.
+2. For each language that the plugin supports, create a translation file.
 
-The structure is similar to `.po` files in that the English string left and its corresponding translation on the right.
+    The files should be JavaScript files and use the relevant language code as the file name. For example: {{site.productname}} will search for a Spanish translation file at `'<your plugin>/langs/es_ES.js'`, where `<your plugin>` is to the directory that contains the plugin's entry point file. For a list of supported languages, see: [Supported languages]({{ site.baseurl }}/configure/localization/#supportedlanguages).
 
-## Example of a Spanish translation for the dialog title
+2. In each translation file, add translation strings by passing an object containing key-value pairs of source strings and translation strings to the [`tinymce.addI18n()` API]({{site.baseurl}}/api/tinymce/root_tinymce/#addi18n).
+3. In the plugin's entry point file, call `tinymce.PluginManager.requireLangPack()` and pass it the plugin identifier and a comma deliminated string of the language codes to load.
```
Comment (tylerkelly13) [inline advanced/creating-a-plugin.md:99]:
```suggestion
3. In the plugin's entry point file, call `tinymce.PluginManager.requireLangPack()` and pass it the plugin identifier and a comma delimitated string of the language codes to load.
```

File: advanced/creating-a-plugin.md (hunk 12 above)
Comment (tylerkelly13) [reply advanced/creating-a-plugin.md:33]: > {{ site.companyname }} recommend including
s/recommend/recommends/
Otherwise, LGTM

**Review by tylerkelly13 — APPROVED** [review]
a few more minor things, then :shipit:

File: advanced/creating-a-plugin.md [hunk 14]
```diff
@@ -7,139 +7,120 @@ description: A short introduction to creating plugins for TinyMCE along with an
...
+{{site.productname}} is designed to be easily extended by custom plugins; with APIs for registering custom plugins, and creating and localizing custom UI.
 
-## File Structure
+## Requirements
 
-First create a directory in the {{site.productname}} plugins directory. {{site.productname}} loads the `plugin.js` file when developers use the `tinymce.js` file in the page. Similarly, the `tinymce.min.js` loads the `plugin.min.js` file. The recommended usage is to use the `tinymce.js` file while developing and then use a build script minifying the `plugin.js` into `plugin.min.js` in production.
+To be recognised as a plugin by {{site.productname}}, the code for your custom plugin must have a JavaScript file with a single entry point that registers the plugin with {{site.productname}} using the `PluginManager` API. Any other code or resources can be in separate files and can be loaded in any standard manner. {{site.productname}} also has various APIs for loading [scripts]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom.scriptloader/) and [stylesheets]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom.stylesheetloader/).
```
Comment (metricjs) [inline advanced/creating-a-plugin.md:14]: Just realised there's a "your" in here...
```suggestion
To be recognised as a plugin by {{site.productname}}, the code for a custom plugin must have a JavaScript file with a single entry point that registers the plugin with {{site.productname}} using the `PluginManager` API. Any other code or resources can be in separate files and can be loaded in any standard manner. {{site.productname}} also has various APIs for loading [scripts]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom.scriptloader/) and [stylesheets]({{ site.baseurl }}/api/tinymce.dom/tinymce.dom.stylesheetloader/).
```

File: advanced/creating-a-plugin.md [hunk 15]
```diff
@@ -7,139 +7,120 @@ description: A short introduction to creating plugins for TinyMCE along with an
...
+```js
+tinymce.init({
+  selector: 'textarea',  // change this value according to your HTML
+  plugins: 'pluginId'
+});
+```
+
+## Testing {{site.productname}}
+
+Due to the range of browser APIs used by {{site.productname}}; when testing {{site.productname}} or any custom plugins, you must use a testing framework that runs on a real browser. Testing frameworks that mock the browser will not work.
```
Comment (metricjs) [inline advanced/creating-a-plugin.md:87]: Another "you"?

File: _includes/live-demos/custom-plugin/index.js [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        name: 'Example plugin',
+        url: 'http://exampleplugindocsurl.com'
+      };
+    }
+  };
+});
+
+/*
+  The following is an example of how to use the new plugin and add the new
+  toolbar button.
```
[the remark this answers is not in the document]
Comment (metricjs) [reply _includes/live-demos/custom-plugin/index.js:68]: Just read weirdly

File: advanced/creating-a-plugin.md (hunk 15 above)
Comment (metricjs) [reply advanced/creating-a-plugin.md:87]:
```suggestion
Due to the range of browser APIs used by {{site.productname}}; when testing {{site.productname}} or any custom plugins, {{site.productname}} requires a testing framework that runs on a real browser. Testing frameworks that mock the browser will not work.
```

**Review by JamesToohey — APPROVED** [review]

**Review by EkimChau — APPROVED** [review]

**Pull request merged**